mad-dog/src/helpers.js

```diff
@@ -260,7 +260,8 @@ const getRandomUser = () => {
     profile_picture_sizes: null,
     cover_photo: null,
     cover_photo_sizes: null,
-    creator_node_endpoint: ''
+    creator_node_endpoint: '',
+    is_deactivated: false
```

Comment:
is this required?

Reply:
It's not listed as a required field, but because the type is `"boolean"` the validator chokes since its type is `undefined` it seems like. This is a quick and dirty change but I think I want to spend more time here tomorrow and investigate, maybe could set the type to `["boolean", "null"]` in the schema?

Reply:
Ah, ironically the schema validator would have replaced the value with the default value if it _was_ required?

https://github.com/AudiusProject/audius-protocol/blob/0a756be79b326a59620e30d484464a2227d7caa2/libs/src/services/schemaValidator/index.js#L54

Reply:
maybe makes sense to do as required and set default value to `false`?

Reply:
i guess fundamentally the question is -- 

do we want all props in the content node metadata blob for a user/track/playlist to be required or optional. it has a material impact on the number of bytes sent across the "wire," but would be far easier to reason with

Reply:
wdyt @csjiang  / @rickyrombo ?

Reply:
Looking into this now - this shouldn't have broken validation, unset properties are supposed to be valid if they're not in the `required` list. Did a simple test and verified that basic schema validation doesn't fail if `is_deactivated` is not included. Trying to figure out why this was breaking then, and why this change fixed it.
```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$ref": "#/definitions/User",
    "definitions": {
        "User": {
            "type": "object",
            "additionalProperties": true,
            "properties": {
                "is_deactivated": {
                    "type": "boolean",
                    "default": false
                }
            },
            "required": [
            ],
            "title": "User"
        }
    }
}

```

```js
audiusLibs.schemas.UserSchema.validate({})
undefined
audiusLibs.schemas.UserSchema.validate({"is_deactivated": undefined})
undefined
audiusLibs.schemas.UserSchema.validate({"is_deactivated": 0 })
index.js:40 Uncaught Error: UserSchema validation failed with errors: [{"property":"instance.is_deactivated","message":"is not of a type(s) boolean","schema":{"type":"boolean","default":false},"name":"type","argument":["boolean"],"stack":"instance.is_deactivated is not of a type(s) boolean"}]
    at Object._this.schemas.<computed>.validate (index.js:40)
    at <anonymous>:1:31
```

Reply:
https://github.com/AudiusProject/audius-protocol/blob/926129262e768b63b4d9fcdda70df0d76212e305/libs/src/api/user.js#L752

we set unset fields to null, which is not the same as undefined, so if I had the type be `[ "boolean", "null" ]` this wouldn't have been invalid.

Reply:
Takeaways/Summary from my investigation:

- We have a schema that lists "type": ["string", "null"] for a majority of types, and has a default of null for those properties, and a schema validator that sets those properties to that default if they’re unset before running validation: https://github.com/AudiusProject/audius-protocol/blob/0a756be79b326a59620e30d484464a2227d7caa2/libs/src/services/schemaValidator/index.js#L54 
  - This doesn’t fit with my understanding of what it means to be “required” - required fields shouldn’t be allowed to be unset and have defaults, they should be always explicitly set. I know these fields need to have defaults/fallbacks etc for things like indexing not to break, but we should revisit how we work around this - maybe by not having them be required in the schema but setting the defaults of non-required fields too? It’s very semantically confusing but I don’t fully grasp all the edge cases here. It definitely seems like we’re overloading “required” (or maybe “validation”) to mean several things
  - Why do we only use the default value on required fields? Shouldn’t we set the defaults for all unset properties?
- In libs, we check for missing fields on user metadata and set them to null: https://github.com/AudiusProject/audius-protocol/blob/926129262e768b63b4d9fcdda70df0d76212e305/libs/src/api/user.js#L752
  - This seems redundant since we already do that for required fields in the validator itself, right? Is there a reason we need this here too?
  - Unlike the validator, this doesn’t use the schema and so this doesn’t use the default field and doesn’t only set required fields - thus we can’t have optional, non-nullable properties in the schema.
- In libs, there’s a separate list of fields for user metadata for things that are required/optional in the JS code: https://github.com/AudiusProject/audius-protocol/blob/926129262e768b63b4d9fcdda70df0d76212e305/libs/src/api/user.js#L9
  - Could we be loading (or if perf is a concern, autogenerating) this from the schemas instead, so that there can be a single source of truth?

Reply:
the USER_METADATA and _cleanUserMetadata predate the schema validator and should totally get axed

Reply:
re point #1: { field: null } can happen when field is required. i think default may be the word that is confusing? i think in the comment you linked default is referring to a non-null value

Reply:
`Why do we only use the default value on required fields? Shouldn’t we set the defaults for all unset properties?`

not including a field on the metadata object is a pattern that we use -- to avoid bloat if not strictly required. it's the undefined vs. null difference. maybe we should drop support for undefined b/c this is confusing